compose-builder/tui-generator.sh

```diff
@@ -0,0 +1,299 @@
+#  *******************************************************************************/
+
+# generator.sh - provide a Terminal User Interface to make generation of different EdgeX Foundry
+# services more comprehensible.
+# Author: Shantanoo (Shan) Desai <shantanoo.desai@gmail.com>
+
+# Map the available options in `make compose` either as standalone strings or array of strings
+
+# Individual Compose Descriptions that may not fall into EdgeX-Foundry Architecture Components
+# Or may be an initial "out-of-the-box" configuration
+
+# Check if whiptail exists to render the UI, if not, exit script
+WHIPTAIL=$(which whiptail)
+
+if [ -z $WHIPTAIL ]
+then
+    echo "This script requires whiptail to render the UI."
+    exit 1;
+fi
+
+MAKE=/usr/bin/make
+
+SELECTED_DEVSERVICES=()
+SELECTED_APPSERVICES=()
+SELECTED_BUS=()
+SELECTED_OTHERS=()
+
+## Additional Options
+additionalOpts=(
+    "mqtt-broker"
+    "modbus-sim"
+)
+
+## General Options Description
+declare -A additionalOptsDesc=(
+    [modbus-sim]="Include ModBus Simulator"
+    [mqtt-broker]="Include MQTT Broker"
+)
+
+
+## Available App Services in EdgeX-Foundry
+appServices=(
+    "asc-http"
+    "asc-mqtt"
+    "asc-sample"
+    "as-llrp"
+    "asc-ex-mqtt"
+)
+
+## App Service Descriptions
+declare -A appServiceDesc=(
+    [asc-http]="Include HTTP Export App Service"
+    [asc-mqtt]="Include MQTT Export App Service"
+    [asc-sample]="Include Sample App Service"
+    [as-llrp]="Include RFID LLRP Inventory"
+    [asc-ex-mqtt]="Include External MQTT Trigger App Service"
+)
+
+
+## Available Device Services in EdgeX-Foundry
+deviceServices=(
+    "ds-bacnet"
+    "ds-camera"
+    "ds-grove"
+    "ds-modbus"
+    "ds-mqtt"
+    "ds-rest"
+    "ds-snmp"
+    "ds-virtual"
+    "ds-coap"
+    "ds-gpio"
+    "ds-llrp"
+)
+
+## Device Service Descriptions
+declare -A deviceServiceDesc=(
+    [ds-bacnet]="Include BACnet Device Service"
+    [ds-camera]="Include Camera Device Service"
+    [ds-grove]="Include Grove Device Service (valid only: arm64)"
+    [ds-modbus]="Include ModBus Device Service"
+    [ds-mqtt]="Include MQTT Device Service"
+    [ds-rest]="Include REST Device Service"
+    [ds-snmp]="Include SNMP Device Service"
+    [ds-virtual]="Include Virtual Device Service"
+    [ds-coap]="Include CoAP Device Service"
+    [ds-gpio]="Include GPIO Device Service"
+    [ds-llrp]="Include RFID LLRP Device Service"
+
+)
+
+
+## Available Message Bus Services in EdgeX-Foundry
+messageBus=(
+    "mqtt-bus"
+    "zmq-bus"
+)
+
+
+## Message Bus Descriptions
+declare -A msgBusDesc=(
+    [mqtt-bus]="Configure MQTT Message Bus"
+    [zmq-bus]="Configure ZMQ Message Bus"
+)
+
+
+####################################################################
+#    Additional Services Options Display Function                  #
+####################################################################
+function additionalServiceOption() {
+    message="Some Additional Services are available that can also be included. \n
```

Comment:
Anywhere you have defined the `message` here like this, you should probably make it a `local` like so:
```shell
local message="...."
```